src/os/bluestore/aio.cc

```diff
@@ -39,18 +98,55 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
 #if defined(HAVE_LIBAIO)
     r = io_submit(ctx, std::min(left, max_iodepth), (struct iocb**)(piocb + done));
 #elif defined(HAVE_POSIXAIO)
+    dout(20) << __func__
+        << " left = " << left
+        << " aio " << &(piocb[done]->aio)
+        << dendl;
     if (piocb[done]->n_aiocb == 1) {
+      dout(25) << __func__
+          << "  n_aiocb == 1 "
+          << " &piocb[done]->aio.aiocbp[0] = " << &(piocb[done]->aio.aiocbp[0])
+          << " on kqueue = " << ctx
+          << " for (fd, off, len) = ("
+            << piocb[done]->aio.aiocbp[0].aio_fildes << ","
+            << piocb[done]->aio.aiocbp[0].aio_offset << ","
+            << piocb[done]->aio.aiocbp[0].aio_nbytes
+          << ")"
+          << dendl;
+      ceph_assert( 0 < piocb[done]->aio.aiocbp[0].aio_fildes
+                   && piocb[done]->aio.aiocbp[0].aio_fildes < 4096 );
       // TODO: consider batching multiple reads together with lio_listio
-      piocb[done]->aio.aiocb.aio_sigevent.sigev_notify = SIGEV_KEVENT;
-      piocb[done]->aio.aiocb.aio_sigevent.sigev_notify_kqueue = ctx;
-      piocb[done]->aio.aiocb.aio_sigevent.sigev_value.sival_ptr = piocb[done];
-      r = aio_read(&piocb[done]->aio.aiocb);
+      piocb[done]->aio.aiocbp[0].aio_sigevent.sigev_notify = SIGEV_KEVENT;
+      piocb[done]->aio.aiocbp[0].aio_sigevent.sigev_notify_kqueue = ctx;
+      piocb[done]->aio.aiocbp[0].aio_sigevent.sigev_notify_kevent_flags =
+                                                                EV_ONESHOT;
+      // make that kevent can read back this piocb
+      piocb[done]->aio.aiocbp[0].aio_sigevent.sigev_value.sival_ptr =
+                                                    (void*)&(piocb[done]->aio);
+      switch (piocb[done]->aio.aiocbp[0].aio_lio_opcode) {
+        case LIO_WRITE:
+          r = aio_write(&(piocb[done]->aio.aiocbp[0]));
+          break;
+        case LIO_READ:
+          r = aio_read(&(piocb[done]->aio.aiocbp[0]));
+          break;
+        default:
```

Comment:
might want to handle `LIO_NOP` as well?

Reply:
@tchaikov 
I do not see any code issue a `LIO_NOP`, so you think it is really needed?

Now it is caugth by the default. I've initialize `r = 0` on every itteration, so it is counted as a successful instruction.

Reply:
just for the sake of completeness. i think we can just list it as another 
```c++
case LIO_NOP:
  break;
```
instead of using `default` to catch it.